### PR TITLE
libc/misc: add lib_glob

### PR DESCRIPTION
**Pull request by FASTSHIFT**
## Summary
Add `glob`.

Reference:
https://man7.org/linux/man-pages/man7/glob.7.html

### CODE REVIEW

File: include/glob.h [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+ * Name: glob
+ *
+ * Description:
+ *   find pathnames matching a pattern.
+ *
+ ****************************************************************************/
+
+int glob(const char *pat, int flags,
+         int (*errfunc)(const char *path, int err),
+         glob_t *g);
```
Comment (gustavonihei) [inline include/glob.h:93]:
```suggestion
int glob(FAR const char *pat, int flags,
         CODE int (*errfunc)(FAR const char *path, int err),
         FAR glob_t *g);
```
This code is common to all architectures, so pointers to data and code need to be defined with the `FAR` and `CODE` qualifiers, respectively.

Please double check the changes in this PR, there are several occurrences.

**Review by gustavonihei — CHANGES_REQUESTED** [review]
Missing qualifiers on pointers to code and data.

File: include/glob.h (hunk 1 above)
Comment (FASTSHIFT) [reply include/glob.h]: Done

File: libs/libc/misc/lib_glob.c [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+    {
+      buf[pos++] = *pat++;
+    }
+
+  /* Consume maximal [escaped-]literal prefix of pattern, copying
+   * and un-escaping it to the running buffer as we go.
+   */
+
+  ptrdiff_t i = 0;
+  ptrdiff_t j = 0;
```
Comment (gustavonihei) [inline libs/libc/misc/lib_glob.c:128]: Another NuttX restriction for generic code is the compliance to C89 standard, so variable declarations must be done in the beginning of the scope.

Comment (FASTSHIFT) [reply libs/libc/misc/lib_glob.c]: Done

**Comment by gustavonihei** [comment]
Do you have any tests for ensuring the implementation is correct? Regarding coding style, the PR seem okay now.

**Review by gustavonihei — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
We test with some open source project, the result is right in our test case.

**Review by gustavonihei — APPROVED** [review]

**Comment by gustavonihei** [comment]
> We test with some open source project, the result is right in our test case.

It would be interesting if you could contribute it and make it part of the CI for `sim` jobs.

**Pull request merged**